### PR TITLE
CMS-354 - Nginx Redirects

### PR DESCRIPTION
**Pull request by mattoberle**

### CODE REVIEW

File: .dockerignore [hunk 1]
```diff
@@ -1,4 +1,8 @@
-tmp
```
Comment (noslouch) [inline .dockerignore]: `tmp` and `dist` are generated by the static ember build, so I think we should keep ignoring them

Comment (mattoberle) [reply .dockerignore]: good catch, I just copied this from another fastboot repo and clobbered what we had there!

File: Dockerfile [hunk 2]
```diff
@@ -1,14 +1,24 @@
 FROM node:8
```
Comment (noslouch) [inline Dockerfile]: maybe let's try it as node:10 if we're updating anyway?

File: nginx/cors.conf [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# vim: syntax=nginx ts=4 sw=4 sts=4 sr noet
```
Comment (noslouch) [inline nginx/cors.conf]: why is there a CORS config? we we expecting API calls to come into the fastboot server?

Comment (mattoberle) [reply nginx/cors.conf]: @noslouch good q, if we don't need it, I'll take it out-- it's just a default pattern I copied from pledge/api-gateway.

File: .dockerignore (hunk 1 above)
Comment (mattoberle) [reply .dockerignore]: re-added.

File: Dockerfile (hunk 2 above)
Comment (mattoberle) [reply Dockerfile]: updated.

File: nginx/cors.conf (hunk 3 above)
Comment (mattoberle) [reply nginx/cors.conf]: removed.

File: nginx/nginx.conf [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+    sendfile        on;
+    tcp_nopush      on;
+    tcp_nodelay     on;
+    keepalive_timeout  130;
+    gzip  on;
+
+    # microcaching
+    # The proxy_cache_path will be used to store responses for 1s on select
+    # endpoints. This improves performance with minimal impact.
+    proxy_cache_path /tmp/cache keys_zone=cache:10m levels=1:2 inactive=600s max_size=100m;
```
Comment (ncastro-nypr) [inline nginx/nginx.conf]: I don't think we need the microcaching here

File: nginx/headers.conf [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Fastboot tries to set a cookie that no one needs... yet,
+# this directive removes it.
+proxy_hide_header "Set-Cookie";
```
Comment (noslouch) [inline nginx/headers.conf]: this is fine, but FYI it's not fastboot per se that sets the cookie, it's a [diff third-party library](https://github.com/simplabs/ember-simple-auth) that's in use on pledge and studios that sets the cookie (for auth purposes).

Hitting _this_ web client as it is, you can see there's no cookie:
```sh
$ curl -si https://gothamist-client.demo.nypr.digital | head

HTTP/2 200
content-type: text/html; charset=utf-8
date: Fri, 12 Apr 2019 17:33:58 GMT
x-powered-by: Express
vary: Accept-Encoding
x-cache: Miss from cloudfront
via: 1.1 a171e7841621158d2aaef33e456c3688.cloudfront.net (CloudFront)
x-amz-cf-id: NwUWGrRBNVwCARuaNc8CflKfm-dOjxse9wichvzWUpo4X3tz0_JU3g==

<!DOCTYPE html>
...
```

File: nginx/nginx.conf (hunk 4 above)
Comment (ncastro-nypr) [reply nginx/nginx.conf]: I guess it doesn't really hurt anything to keep it

File: nginx/mime.types [hunk 6]
```diff
@@ -0,0 +1,89 @@
+
```
Comment (noslouch) [inline nginx/mime.types]: all these mime types are probably a best practice, but we're only expecting the fastboot server to serve HTML files; everything else should be a direct request to the asset via cloudfront or an API call on a different server.

See this PR for more context: https://github.com/nypublicradio/nypr-fastboot/pull/2

File: nginx/nginx.conf (hunk 4 above)
Comment (mattoberle) [reply nginx/nginx.conf]: Yeah it probably won't be hit much, but I kept this in as a kind of low effort rate-limiting measure (if say, a thundering was let through to the backend). Maybe that's not a concern with CloudFront.

File: nginx/nginx.conf [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+error_log  /dev/stdout warn;
+env ENV;
+
+events {
+    worker_connections  4096;
+}
+
+http {
+    include       mime.types;
+    default_type  application/octet-stream;
```
Comment (noslouch) [inline nginx/nginx.conf]: if you're not using the mime types module, then maybe just set html as the default here

Comment (mattoberle) [reply nginx/nginx.conf]: done

File: nginx/mime.types (hunk 6 above)
Comment (mattoberle) [reply nginx/mime.types]: removed in favor of just using whatever nginx provides by default

File: nginx/headers.conf (hunk 5 above)
Comment (mattoberle) [reply nginx/headers.conf]: Per discussion on Slack-- leaving this in as a cruel way to force re-evaluation of cache rules if a cookie-setting library is added later.

File: nginx/nginx.conf [hunk 8]
```diff
@@ -0,0 +1,116 @@
...
+        location = /cat.php { return 301 /$arg_term/$arg_page; }
+        location = /getaways.php { return 301 /getaways/$arg_page; }
+        location = /mediakit/gothamist.php { return 301 https://www.gothamistllc.com/gothamist; }
+        location = /mediakit/index.php { return 301 https://www.gothamistllc.com; }
+        location = /mediakit/spec_rates.php { return 301 https://www.gothamistllc.com/offerings/; }
+        location = /mobile.php { return 301 /$arg_page; }
+        location = /pretty.php { return 301 /; }
+        location = /tag.php { return 301 /tags/$arg_term/$arg_page; }
+
+        location = /archives/food.xml { return 301 /food/index.xml; }
```
Comment (ncastro-nypr) [inline nginx/nginx.conf]: I'm not sure if this will resolve on the new site, it'll be something to look out for

File: nginx/nginx.conf [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+        location = /labs/map { return 301 /map; }
+        location = /submit { return 301 https://gothamist.submittable.com/submit/18730; }
+        location = /turnout { return 301 https://project.gothamist.com/voter-turnout; }
+
+        location = /atom.xml { return 301 https://feeds.gothamistllc.com/gothamist05; }
+        location = /index.rdf { return 301 https://feeds.gothamistllc.com/gothamist05; }
+        location = /index.xml { return 301 https://feeds.gothamistllc.com/gothamist05; }
+
+        location ~ ^/archives/(.*)\.(php|xml)$ { return 301 /$1.$2; }
+        location ~^/blog[2-6]?\.php$ { return 301 /; }
```
Comment (ncastro-nypr) [inline nginx/nginx.conf]: missing a space here I think

File: nginx/nginx.conf (hunk 8 above)
Comment (mattoberle) [reply nginx/nginx.conf]: good point.
what are the XML pages for anyway, do you know @ncastro-nypr ?
something w/ their automated newsletters?

Comment (mattoberle) [reply nginx/nginx.conf]: marked this as a TODO, pending slack answers.

**Pull request merged**